### PR TITLE
[REVIEW] Add DateTimeIndex.weekday

### PR DESCRIPTION
**Pull request by brandon-b-miller**
Small implementation of https://github.com/rapidsai/cudf/issues/2435, needs a test in the cpp layer still. 

uses [Zeller's algorithm](https://en.wikipedia.org/wiki/Zeller%27s_congruence)

### CODE REVIEW

File: cpp/src/datetime/datetime_ops.cu [hunk 1]
```diff
@@ -442,6 +534,50 @@ gdf_error gdf_extract_datetime_day(gdf_column *input, gdf_column *output) {
...
+
+	GDF_REQUIRE(input->size == output->size, GDF_COLUMN_SIZE_MISMATCH);
+	GDF_REQUIRE(output->dtype == GDF_INT16, GDF_UNSUPPORTED_DTYPE);
+
+	cudaStream_t stream;
+	cudaStreamCreate(&stream);
+
+
+    if (input->valid){
+      gdf_size_type num_bitmask_elements = ( ( input->size +( GDF_VALID_BITSIZE - 1)) / GDF_VALID_BITSIZE );
```
Comment (brandon-b-miller) [inline cpp/src/datetime/datetime_ops.cu]: I am not actually sure what this line of code does, I just used the other fields as a template essentially. So let me know if this is wrong.

File: cpp/include/cudf/functions.h [hunk 2]
```diff
@@ -362,6 +362,16 @@ gdf_error gdf_extract_datetime_month(gdf_column *input, gdf_column *output);
...
 
+/**
+ * @brief  Extracts day from any date time type and places results into a preallocated GDF_INT16 column
+ *
+ * @param[in] gdf_column of the input
+ * @param[out] output gdf_column. The output memory needs to be preallocated
+ *
+ * @returns GDF_SUCCESS upon successful compute, otherwise returns appropriate error code
+ */
+gdf_error gdf_extract_datetime_weekday(gdf_column *input, gdf_column *output);
```
Comment (jrhemstad) [inline cpp/include/cudf/functions.h]: We no longer put APIs in `functions.h`. See https://github.com/rapidsai/cudf/blob/branch-0.10/cpp/docs/TRANSITIONGUIDE.md for the guide on API design. 

Suggested new API:
```
namespace cudf{
gdf_column extract_datetime_weekday(gdf_column const& input);
}
```

File: cpp/src/datetime/datetime_ops.cu [hunk 3]
```diff
@@ -442,6 +534,50 @@ gdf_error gdf_extract_datetime_day(gdf_column *input, gdf_column *output) {
...
+      thrust::copy(rmm::exec_policy(stream)->on(stream), input->valid, input->valid + num_bitmask_elements, output->valid); // copy over valid bitmask
+    }
+	if ( input->dtype == GDF_DATE64 ) {
```
Comment (jrhemstad) [inline cpp/src/datetime/datetime_ops.cu]: This needs to use the [`type_dispatcher`](https://github.com/rapidsai/cudf/blob/branch-0.10/cpp/include/cudf/utilities/legacy/type_dispatcher.hpp) with a specialization for the timestamp/date types.

File: cpp/src/datetime/datetime_ops.cu [hunk 4]
```diff
@@ -442,6 +534,50 @@ gdf_error gdf_extract_datetime_day(gdf_column *input, gdf_column *output) {
 	return GDF_SUCCESS;
 }
 
+gdf_error gdf_extract_datetime_weekday(gdf_column *input, gdf_column *output) {
+
+	GDF_REQUIRE(input->size == output->size, GDF_COLUMN_SIZE_MISMATCH);
+	GDF_REQUIRE(output->dtype == GDF_INT16, GDF_UNSUPPORTED_DTYPE);
+
+	cudaStream_t stream;
+	cudaStreamCreate(&stream);
```
Comment (jrhemstad) [inline cpp/src/datetime/datetime_ops.cu]: Don't create streams, it's expensive.

File: cpp/src/datetime/datetime_ops.cu [hunk 5]
```diff
@@ -442,6 +534,50 @@ gdf_error gdf_extract_datetime_day(gdf_column *input, gdf_column *output) {
 	return GDF_SUCCESS;
 }
 
+gdf_error gdf_extract_datetime_weekday(gdf_column *input, gdf_column *output) {
+
+	GDF_REQUIRE(input->size == output->size, GDF_COLUMN_SIZE_MISMATCH);
```
Comment (jrhemstad) [inline cpp/src/datetime/datetime_ops.cu]: Use `CUDF_EXPECTS`.

File: cpp/src/datetime/datetime_ops.cu (hunk 1 above)
Comment (jrhemstad) [reply cpp/src/datetime/datetime_ops.cu]: Yikes, code like this shouldn't still exist :crying_cat_face: 

Use https://github.com/rapidsai/cudf/blob/12f15c81c36659cdec2935cfec38b02968a4deda/cpp/include/cudf/functions.h#L76 instead.

File: cpp/src/datetime/datetime_ops.cu (hunk 3 above)
Comment (jrhemstad) [reply cpp/src/datetime/datetime_ops.cu]: You can make `gdf_extract_weekday_from_unixtime_op` be a type-dispatched function object.

File: cpp/src/datetime/datetime_ops.cu (hunk 1 above)
Comment (brandon-b-miller) [reply cpp/src/datetime/datetime_ops.cu]: fixed.

File: cpp/src/datetime/datetime_ops.cu (hunk 5 above)
Comment (brandon-b-miller) [reply cpp/src/datetime/datetime_ops.cu]: Fixed (I think.)

File: cpp/include/cudf/functions.h (hunk 2 above)
Comment (brandon-b-miller) [reply cpp/include/cudf/functions.h]: fixed, let me know if this is what you meant.

**Comment by brandon-b-miller** [comment]
@harrism Indeed that is what this is, admittedly I took this issue on somewhat assuming I could just use the other fields as a template without being all that familiar with the c++ layer. That said I understand the motivation for the changes that @jrhemstad already requested here, such as adhering to the new API standards, etc. 

I don't think it would take too much work to apply some of these suggestions across the entirety of datetime_ops.cu, such as migrating the API out of functions.h, and I would be happy to do so. We could even consider a datetime_ops.cu refactor as "blocking" this. However I have I think the same sense as you that doing so doesn't quite seem like it falls within the original scope.

**Comment by harrism** [comment]
I'll file an issue. Don't do it in this PR.

**Comment by brandon-b-miller** [comment]
There is one small item that needs reviewing (the implementation of values_host) and other then that I believe this is ready to merge, should be in within the next day or so.

**Comment by brandon-b-miller** [comment]
rerun tests

**Comment by brandon-b-miller** [comment]
rerun tests

File: cpp/src/datetime/datetime_ops.cu (hunk 5 above)
Comment (jrhemstad) [reply cpp/src/datetime/datetime_ops.cu]: This hasn't changed?

File: cpp/src/datetime/datetime_ops.cu (hunk 4 above)
Comment (jrhemstad) [reply cpp/src/datetime/datetime_ops.cu]: ?

**Comment by jrhemstad** [comment]
So did we decide we aren't making any changes to the API and such?

**Comment by brandon-b-miller** [comment]
@jrhemstad I worked through some of the suggestions you made in a few of the previous commits, but ended up [reverting them],(https://github.com/rapidsai/cudf/pull/2814/commits/cc522e361ca461a570b01f1ca0a93476ce48157b) which is why things look unchanged since last we last looked. My understanding is that since the issues that were noted are broadly applicable to all the datetime ops in `datetime_ops.cu` (week, month, year, day, etc), it makes sense to work on a refactor of all of the functions in a separate PR (I would be happy to do this).

**Comment by brandon-b-miller** [comment]
rerun tests

**Comment by harrism** [comment]
Let's go ahead. Happy to have your contribution on the refactor! 

Unfortunately CI is failing here.

**Comment by brandon-b-miller** [comment]
Looks like CI issues are due to an unrelated conda discrepancy, I believe a fix is in flight - should be ready to go soon.

**Comment by beckernick** [comment]
It would be nice to define dayofweek as an alias for weekday to match pandas, too.

https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#time-date-components

**Comment by brandon-b-miller** [comment]
rerun tests

**Comment by brandon-b-miller** [comment]
rerun tests

**Comment by harrism** [comment]
rerun tests

**Comment by brandon-b-miller** [comment]
rerun tests

**Pull request merged**